### PR TITLE
Produce Publication (doi) DTOs on Event Bridge

### PR DESCRIPTION
**Pull request by norrs**
* DynamodbEvent Stream attached to emit it's events to Event Bridge. It emits list of event bridge DynamodbEvent.StreamRecords (decouples from dynamodb)
* Consumer of Event Bridge DynamodbEvent.StreamRecord which produces the (doi) Publication DTO from DynamodbEvent.StreamRecord (1 fan out. Search can use the same EventBridge source from bullet point above))
* Configured manually the `rono-sqs` to be consumer with EventBridge rules to verify it worked:

```
{
  "source": ["lambda"],
  "resources": [
    "arn:aws:events:eu-west-1:884807050265:event-bus/default"
  ]
}
```
(ie all messages on our event-bus)

```
{
  "source": [
    "lambda"
  ],
  "detail": {
    "responsePayload": {
      "type": [
        "doi.publication"
      ]
    }
  }
}
```
(all messages from our fan-out , this is the event rule type we want to use!)

Both works. 

Example above should also add `detail-type` with the missing `"lambda success output from lambda destination"`

### CODE REVIEW

**Comment by norrs** [comment]
This is WIP and pair programming between @axthosarouris , @janbj and @norrs .

**Comment by norrs** [comment]
This branch currently has automatic deployment with stack `rono-publication` something :)

File: publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java [hunk 1]
```diff
@@ -19,10 +22,13 @@
 import no.unit.nva.publication.doi.dynamodb.dao.DynamodbStreamRecordJsonPointers.DynamodbImageType;
 import nva.commons.utils.JsonUtils;
 
+import static java.util.function.Predicate.not;
+
 public class PublicationMapper {
```
Comment (axthosarouris) [inline publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java:27]: Add a short comment about the functionality of this class

File: publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java [hunk 2]
```diff
@@ -49,7 +55,7 @@ private static URI transformIdentifierToId(String namespace, String identifier)
      * Map a DynamodbStreamRecord with oldImage and/or newImage to PublicationMapping. Publication is a wrapper object
      * containing mapped old and/or new Publication.
      *
-     * @param streamRecord  DynamodbStreamRecord
+     * @param streamRecord DynamodbStreamRecord
      * @return PublicationMapping
      */
     public PublicationMapping fromDynamodbStreamRecord(DynamodbStreamRecord streamRecord) {
```
Comment (norrs) [inline publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java:61]: note to self:

This method doesnt belong in this class I believe @janbj , this one should use our DAOs.

Only DAOs should access the DynamodbStreamRecord .

File: publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbEventConducer.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                                                 AwsEventBridgeEvent<DynamodbEvent> event,
+                                                 Context context) {
+        return fromDynamodbStreamRecords(input.getRecords());
+    }
+
+    private PublicationCollection fromDynamodbStreamRecords(List<DynamodbEvent.DynamodbStreamRecord> records) {
+        List<Publication> dtos = new ArrayList<>();
+        for (DynamodbEvent.DynamodbStreamRecord record : records) {
+            var publicationMapping = publicationMapper.fromDynamodbStreamRecord(record);
+            if (isEffectiveChange(publicationMapping) && publicationMapping.getNewPublication().isPresent()) {
```
Comment (axthosarouris) [inline publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbEventConducer.java:50]: why are you checking twice that newPublicationExists? You are checking this already in isEffectiveChange.  If you want to get rid of the warning do not use the `Optional::get` but the `Optional:orElseThrow`. As Brian Goetz says `get ` was a mistake. 
You can also use streams to parallelize the processing.

```
 List<Publication> dtos = records.stream()
            .parallel() // optional step.
            .map(publicationMapper::fromDynamodbStreamRecord)
            .filter(this::isEffectiveChange)
            .map(publicationMapping -> publicationMapping.getNewPublication().orElseThrow())
            .collect(Collectors.toList());

        logger.info("From {} records we made {} Publication DTOs", records.size(), dtos.size());
        return new PublicationCollection(TYPE_DTO_DOI_PUBLICATION, dtos);
```

https://stackoverflow.com/questions/49158806/optional-get-vs-overloaded-optional-orelsethrow-in-java-10

File: publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbEventConducer.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                dtos.add(publicationMapping.getNewPublication().get());
+            }
+        }
+        logger.info("From {} records we made {} Publication DTOs", records.size(), dtos.size());
+        return new PublicationCollection(TYPE_DTO_DOI_PUBLICATION, dtos);
+    }
+
+    private boolean isEffectiveChange(PublicationMapping publicationMapping) {
+        Optional<Publication> newPublication = publicationMapping.getNewPublication();
```
Comment (axthosarouris) [inline publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbEventConducer.java:59]: isn't this equivalent to the following? I think here the usage of nulls simplifies the code.

```
   Publication newPublication = publicationMapping.getNewPublication().orElse(null);
    Publication oldPublication = publicationMapping.getOldPublication().orElse(null);

        if (newPublication != null) {
            return !newPublication.equals(oldPublication);
        }
        return false;
```

File: publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+     * Setting up test environment.
+     */
+    @BeforeEach
+    public void setUp() {
+        handler = new DynamoDbEventConducer(EXAMPLE_NAMESPACE);
+        context = mock(Context.class);
+    }
+
+    @Test
+    void processInput() throws IOException {
```
Comment (axthosarouris) [inline publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java:46]: naming

File: publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        var eventBridgeEvent = new EventParser<DynamodbEvent>(
+            eventFile).parse(DynamodbEvent.class);
+        var actual = handler.processInput(event, eventBridgeEvent, context);
+
+        assertThat(actual.getType(), is(equalTo(DOI_PUBLICATION_TYPE)));
+        assertThat(actual.getItems(), hasSize(1));
+    }
+
+    @Test
+    void noNewImageWhileProcessingInputThenInEffectiveChangeIsNotIncluded() throws JsonProcessingException {
```
Comment (axthosarouris) [inline publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java:58]: does this mean processInputFiltersOutDynamoStreamEventsWithNoNewImage?

File: publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        var eventBridgeEvent = new EventParser<DynamodbEvent>(
+            eventFile).parse(DynamodbEvent.class);
+        var actual = handler.processInput(event, eventBridgeEvent, context);
+
+        assertThat(actual.getType(), is(equalTo(DOI_PUBLICATION_TYPE)));
+        assertThat(actual.getItems(), hasSize(0));
+    }
+
+    @Test
+    void differentOldAndNewImageWhileProcessingInputThenEffectiveChangeIsIncluded() throws JsonProcessingException {
```
Comment (axthosarouris) [inline publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java:70]: processInputIncludesDynamoStreamEventsWhenOldAndNewImageAreDifferent

File: publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        var eventBridgeEvent = new EventParser<DynamodbEvent>(
+            eventFile).parse(DynamodbEvent.class);
+        var actual = handler.processInput(event, eventBridgeEvent, context);
+
+        assertThat(actual.getType(), is(equalTo(DOI_PUBLICATION_TYPE)));
+        assertThat(actual.getItems(), hasSize(1));
+    }
+
+    @Test
+    void equalOldAndNewImageWhileProcessingInputThenInEffectiveChangeIsNotIncluded() throws JsonProcessingException {
```
Comment (axthosarouris) [inline publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java:82]: processInputFiltersOutDynamoStreamEventsWhenOldAndNewImageAreEqual

File: template.yaml [hunk 9]
```diff
@@ -459,11 +462,89 @@ Resources:
...
+                Action:
+                  - events:*
+                  - sqs:SendMessage
+                  - sqs:ReceiveMessage
+                  - sqs:DeleteMessage
+                  - sqs:*
+                  - lambda:InvokeFunction
+                Resource: "*"
+
+  FailQueue:
```
Comment (axthosarouris) [inline template.yaml:503]: General comment: Is this queue going to be used by other handlers outside this stack

**Review by axthosarouris — CHANGES_REQUESTED** [review]

File: publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java [hunk 10]
```diff
@@ -63,20 +69,36 @@ public PublicationMapping fromDynamodbStreamRecord(DynamodbStreamRecord streamRe
 
             if (acceptStreamViewTypes(streamViewType,
                 StreamViewType.NEW_AND_OLD_IMAGES, StreamViewType.OLD_IMAGE)) {
-                Publication oldPublication = fromDynamodbStreamRecordImage(dynamodb.getOldImage());
-                publicationMappingBuilder.withOldPublication(oldPublication);
+                maybeSetOldPublication(publicationMappingBuilder, dynamodb);
```
[the remark this answers is not in the document]
Comment (norrs) [reply publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java:72]: `optionalSetOldPublication`  . `Maybe` is kotlins naming/version of `Optional`. The `if` check is needed because the mapper will throw unrecognized data illegal state exception if not. 

Unsure if we want data classes to have optional getters?

File: publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java (hunk 6 above)
Comment (norrs) [reply publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java:58]: Given `noNewImage` While `processingInput` Then `InEffectiveChangeIsNotIncluded` . 

Or do you want:

While `processingInput` Then `InEffectiveChangeIsNotIncluded` given `NoNewImage` ?

File: template.yaml (hunk 9 above)
Comment (norrs) [reply template.yaml:503]: I think we want to look into stacks should expose resources so others can depend/referere to em. 

But the outputs needs to be joined with the stack name.

File: template.yaml [hunk 11]
```diff
@@ -504,7 +585,7 @@ Resources:
 Outputs:
   FanoutLambdaName:
```
Comment (norrs) [inline template.yaml:586]: `DynamodbEventEventBridgeHandlerName` ?

File: template.yaml [hunk 12]
```diff
@@ -504,7 +585,7 @@ Resources:
...
     Description: "Fanout Lambda Function Name"
```
Comment (norrs) [inline template.yaml:587]: `DynamodbEventEventBridgeHandler Lambda Function Name`?

File: template.yaml [hunk 13]
```diff
@@ -504,7 +585,7 @@ Resources:
...
-    Value: !Ref FanoutLambda
+    Value: !Ref DynamodbEventEventBridgeHandler
   FanoutUrl:
```
Comment (norrs) [inline template.yaml:589]: @janbj Can you describe this? :D

File: publication-doi-commons/src/test/java/no/unit/nva/publication/doi/dto/PublicationCollectionTest.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+class PublicationCollectionTest {
+
+    private static final String EXAMPLE_TYPE = "exampleType";
+    public static final PublicationDynamoEventTestDataGenerator PUBLICATION_GENERATOR =
+        new PublicationDynamoEventTestDataGenerator();
+    private static final String EXAMPLE_NAMESPACE = "http://example.net/namespace/";
+
+    @Test
+    void getType() {
```
[the remark this answers is not in the document]
Comment (norrs) [reply publication-doi-commons/src/test/java/no/unit/nva/publication/doi/dto/PublicationCollectionTest.java]: Fixed.

File: publication-doi-commons/src/test/java/no/unit/nva/publication/doi/dto/PublicationCollectionTest.java [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+        assertThat(new PublicationCollection(EXAMPLE_TYPE, Collections.emptyList()).getType(),
+            is(equalTo(EXAMPLE_TYPE)));
+    }
+
+    @Test
+    void getItems() {
```
[the remark this answers is not in the document]
Comment (norrs) [reply publication-doi-commons/src/test/java/no/unit/nva/publication/doi/dto/PublicationCollectionTest.java]: Fixed.

File: publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java (hunk 5 above)
Comment (norrs) [reply publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java]: Duplicate test, removed.

File: publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java [hunk 16]
```diff
@@ -33,12 +39,12 @@
      * @param namespace Namespace to use for constructing ids from identifiers that are owned by Publication.
      */
     public PublicationMapper(String namespace) {
-        if (namespace == null || !namespace.endsWith("/")) {
+        if (namespace == null || !namespace.endsWith(FORWARD_SLASH)) {
             throw new IllegalArgumentException(ERROR_NAMESPACE_MUST_CONTAIN_SUFFIX_SLASH);
         }
 
         var ns = namespace.toLowerCase(Locale.US);
```
[the remark this answers is not in the document]
Comment (norrs) [reply publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java]: Fixed, inlined it. No need for extra variable.

File: publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java (hunk 6 above)
Comment (axthosarouris) [reply publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java:41]: For readability and for better understanding of the intention of the test we are trying to follow the following convetion:
 methodName(Returns/Throws/Does)SomethingWhen(Input/Condition)IsSomething. 

"While" here is confusing because it has the meaning of "during". We do not care what the function does while it is running. We care about the output. Also, if we do not follow the same convention, then the test becomes harder to understand. 

The target is that the method name should as close to a real sentence as possible. For example, 
"processInput filters out dynamo-stream-events with no new image" is clear about the function that is testing (processInput), the input (dynamo-stream-events) and what it does with them (filters out)


"no new Image while processing input then in effective change is not included"  is not clear about the input (the input is not a new image) the name is implied, and the notion of an "ineffective change" is defined in your code. Basically, It not clear that we are talking about a dynamo stream event and that the ineffective change is the fact that there is no new image. 
In other words one has to read the code, spot the method ineffective change, understand what is happening there and come back here to read the test.  The test however tests something very simple: Are dynamo-stream-events with no new image excluded or not?

Comment (norrs) [reply publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java:41]: Thanks for sharing your thoughts, will update. Makes it easier to understand how you want em written. :)

File: publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java (hunk 1 above)
Comment (norrs) [reply publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java:34]: Added

File: publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (axthosarouris) [reply publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java:78]: After looking around I would name it just `setOldPublication`, you can keep maybe, but I find it confusing. Also if Maybe refers to Optionals, you do not use Optionals.

File: template.yaml (hunk 13 above)
Comment (janbj) [reply template.yaml:589]: Describe how?

File: template.yaml (hunk 11 above)
Comment (janbj) [reply template.yaml:586]: Sure.

File: publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java (hunk 2 above)
Comment (norrs) [reply publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java:67]: We treat `DynamodbEvent` and `DynamodbEvent.StreamRecord` as our own data classes even if it comes from AWS library.

File: publication-event-dtopublicationdoi-producer/build.gradle [hunk 17]
```diff
@@ -0,0 +1,23 @@
+def log4jVersion = "2.13.2"
+
+dependencies {
+    implementation project(":publication-doi-commons")
+
+    implementation group: 'com.github.BIBSYSDEV.nva-datacite-mds', name: 'event-handler-template', version: '087d568dda7adf671c843d5c0294a4d7ff54e756'
```
Comment (janbj) [inline publication-event-dtopublicationdoi-producer/build.gradle:6]: Any releases for this yet?

File: publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbFanoutPublicationDtoProducer.java [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Override
+    protected PublicationHolder processInput(DynamodbEvent.DynamodbStreamRecord input,
+                                             AwsEventBridgeEvent<DynamodbEvent.DynamodbStreamRecord> event,
+                                             Context context) {
+        return fromDynamodbStreamRecords(input);
+    }
+
+    private PublicationHolder fromDynamodbStreamRecords(DynamodbEvent.DynamodbStreamRecord record) {
+        var dto = Optional.ofNullable(record)
```
Comment (janbj) [inline publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbFanoutPublicationDtoProducer.java:45]: Extract to method

File: publication-event-dtopublicationdoi-producer/build.gradle (hunk 17 above)
Comment (norrs) [reply publication-event-dtopublicationdoi-producer/build.gradle:6]: Unsure. We can probably make one? @axthosarouris

File: publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbFanoutPublicationDtoProducer.java [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+            .map(publicationMapper::fromDynamodbStreamRecord)
+            .filter(this::isEffectiveChange)
+            .map(publicationMapping -> publicationMapping.getNewPublication().orElseThrow());
+
+        logger.info("{} Publication DTO from DynamodbStreamRecord", dto.isPresent() ? "Created" : "Skipped creating");
+        return dto
```
Comment (janbj) [inline publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbFanoutPublicationDtoProducer.java:51]: Less important than previous, but maybe extract this one to method too

File: publication-event-emitter/src/test/java/no/unit/nva/doi/lambda/DynamodbEventFanoutStreamRecordsEventBridgeHandlerTest.java [hunk 20]
```diff
@@ -24,11 +30,22 @@ public void handleRequestWritesToConsoleOnValidEvent() throws Exception {
...
         File eventFile = new File(DYNAMODB_STREAM_EVENT);
         DynamodbEvent event = objectMapper.readValue(eventFile, DynamodbEvent.class);
 
         handler.handleRequest(event, context);
     }
+
+    @BeforeEach
+    void setUp() {
+        environmentMock = mock(Environment.class);
+        when(environmentMock.readEnv(ALLOWED_ORIGIN_ENV)).thenReturn(WILDCARD);
```
Comment (janbj) [inline publication-event-emitter/src/test/java/no/unit/nva/doi/lambda/DynamodbEventFanoutStreamRecordsEventBridgeHandlerTest.java:45]: Why do we need this? Do we expose handler in API Gateway?

**Review by janbj — APPROVED** [review]

File: publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbFanoutPublicationDtoProducer.java (hunk 18 above)
Comment (norrs) [reply publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbFanoutPublicationDtoProducer.java]: Fixed.

File: publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbFanoutPublicationDtoProducer.java (hunk 19 above)
Comment (norrs) [reply publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbFanoutPublicationDtoProducer.java:50]: Fixed.

File: publication-event-emitter/src/test/java/no/unit/nva/doi/lambda/DynamodbEventFanoutStreamRecordsEventBridgeHandlerTest.java (hunk 20 above)
Comment (norrs) [reply publication-event-emitter/src/test/java/no/unit/nva/doi/lambda/DynamodbEventFanoutStreamRecordsEventBridgeHandlerTest.java]: Fixed.

File: publication-event-dtopublicationdoi-producer/build.gradle (hunk 17 above)
Comment (norrs) [reply publication-event-dtopublicationdoi-producer/build.gradle]: Fixed by using a pre-release.

**Comment by norrs** [comment]
@axthosarouris @janbj : Please resolve conversation if you agree with fixes. PTAL.

File: publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (norrs) [reply publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java]: Fixed by refactoring.

File: publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java [hunk 21]
```diff
@@ -63,20 +69,36 @@ public PublicationMapping fromDynamodbStreamRecord(DynamodbStreamRecord streamRe
...
-                Publication oldPublication = fromDynamodbStreamRecordImage(dynamodb.getOldImage());
-                publicationMappingBuilder.withOldPublication(oldPublication);
+                maybeSetOldPublication(publicationMappingBuilder, dynamodb);
             }
 
             if (acceptStreamViewTypes(streamViewType,
                 StreamViewType.NEW_AND_OLD_IMAGES, StreamViewType.NEW_IMAGE)) {
-                var newPublication = fromDynamodbStreamRecordImage(dynamodb.getNewImage());
-                publicationMappingBuilder.withNewPublication(newPublication);
+                maybeSetNewPublication(publicationMappingBuilder, dynamodb);
```
[the remark this answers is not in the document]
Comment (norrs) [reply publication-doi-commons/src/main/java/no/unit/nva/publication/doi/PublicationMapper.java]: ditto

File: publication-doi-commons/src/test/java/no/unit/nva/publication/doi/dto/PublicationTest.java [hunk 22]
```diff
@@ -22,7 +22,7 @@
 
     @Test
     public void testBuilder() {
```
[the remark this answers is not in the document]
Comment (norrs) [reply publication-doi-commons/src/test/java/no/unit/nva/publication/doi/dto/PublicationTest.java]: Fixed.

File: publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/AppConfig.java [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+    private AppConfig() {
+    }
+
+    @JacocoGenerated
+    public static String getNamespace() {
+        return extractNamespaceUriFromEnvironment().toString();
+    }
+
+    private static URI extractNamespaceUriFromEnvironment() {
+        // Example values: "CustomDomain": "api.dev.nva.aws.unit.no", "CustomDomainBasePath
```
[the remark this answers is not in the document]
Comment (norrs) [reply publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/AppConfig.java]: Fixed

File: publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbEventConducer.java (hunk 3 above)
Comment (norrs) [reply publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbEventConducer.java]: Thanks, updated.

File: publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbEventConducer.java (hunk 4 above)
Comment (norrs) [reply publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbEventConducer.java]: Updated.

File: publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java (hunk 7 above)
Comment (norrs) [reply publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java]: Updated.

File: publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java (hunk 8 above)
Comment (norrs) [reply publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbEventConducerTest.java]: Updated.

File: publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbFanoutPublicationDtoProducerTest.java [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+        handler.handleRequest(eventInputStream, outputStream, context);
+        PublicationHolder actual = outputToPublicationHolder(outputStream);
+
+        assertThat(actual.getType(), is(equalTo(DOI_PUBLICATION_TYPE)));
+        assertThat(actual.getItem(), notNullValue());
+    }
+
+    @Test
+    void processInputSkipsCreatingDtosWhenNoNewImageIsPresentInDao() throws JsonProcessingException {
+        var eventFile = IoUtils.stringFromResources((DYNAMODB_STREAM_EVENT_OLD_ONLY));
```
Comment (axthosarouris) [inline publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbFanoutPublicationDtoProducerTest.java:60]: fix this test as well as we did with the others.

**Review by axthosarouris — CHANGES_REQUESTED** [review]

**Review by janbj — DISMISSED** [review]

File: publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbFanoutPublicationDtoProducer.java [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+    private PublicationHolder fromDynamodbStreamRecords(DynamodbEvent.DynamodbStreamRecord record) {
+        var dto = mapToPublicationDto(record);
+        logMappingResults(dto);
+        return dto
+            .map(publication -> new PublicationHolder(TYPE_DTO_DOI_PUBLICATION, publication))
+            .orElse(NO_OUTPUT_NO_EVENT);
+    }
+
+    private void logMappingResults(Optional<Publication> dto) {
+        logger.info("{} Publication DTO from DynamodbStreamRecord", dto.isPresent() ? CREATED : SKIPPED_CREATING);
```
[the remark this answers is not in the document]
Comment (norrs) [reply publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbFanoutPublicationDtoProducer.java:56]: With the little traffic we have, I don't see any cost issues. Lets change in to `debug` down the line once it has been running stable for a while.

[the remark this answers is not in the document]
Comment (norrs) [reply publication-event-dtopublicationdoi-producer/src/main/java/no/unit/nva/doi/event/producer/DynamoDbFanoutPublicationDtoProducer.java:56]: And I would imagine you are thinking about the costs about CloudWatch. 

But we don't have Kibana with Elasticsearch backend (for example) to replace it, and we should avoid it and not add it before there is a need. Server less thinking, use what is provided in the platform.

File: publication-event-dtopublicationdoi-producer/build.gradle [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+
+    implementation(group: 'com.github.BIBSYSDEV.nva-datacite-mds', name: 'event-handler-template', version: '0.0.1') {
+        because 'first usage of event-handler-template framework/library, it is high candidate to be moved into its own repository if the first test has been a success.'
+    }
+
+    implementation group: 'com.github.BIBSYSDEV', name: 'nva-commons', version: project.ext.nvaCommonsVersion
+
+    implementation group: 'com.amazonaws', name: 'aws-lambda-java-core', version: '1.2.1'
+    implementation group: 'com.amazonaws', name: 'aws-lambda-java-events', version: '3.2.0'
+    implementation group: 'software.amazon.awssdk', name: 'eventbridge', version: project.ext.awsSdk2Version
```
[the remark this answers is not in the document]
Comment (norrs) [reply publication-event-dtopublicationdoi-producer/build.gradle:14]: The reason why the 2 first ones are included is to get the `DynamodbEvent` and `DynamodbEvent.StreamRecord` POJOs that are defined by amazon.

File: publication-doi-commons/src/main/java/no/unit/nva/publication/doi/dto/PublicationHolder.java [hunk 27]
```diff
@@ -0,0 +1,27 @@
...
+
+public class PublicationHolder {
+    private final String type;
+    private final Publication item;
+
+    @JacocoGenerated
+    @JsonCreator
+    public PublicationHolder(
+            @JsonProperty("type") String type,
+            @JsonProperty("item") Publication publicationList) {
```
[the remark this answers is not in the document]
Comment (norrs) [reply publication-doi-commons/src/main/java/no/unit/nva/publication/doi/dto/PublicationHolder.java]: Fixed, it is a `publication` , parameter name left unchanged from refactor earlier when we changed to emitting from `DynamodbEvent` to `DynamodbEvent.StreamRecord`

File: publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbFanoutPublicationDtoProducerTest.java (hunk 24 above)
Comment (norrs) [reply publication-event-dtopublicationdoi-producer/src/test/java/no/unit/nva/doi/event/producer/DynamoDbFanoutPublicationDtoProducerTest.java]: Fixed

**Review by axthosarouris — APPROVED** [review]

**Review by janbj — APPROVED** [review]
